### PR TITLE
Chat perf part 2

### PR DESCRIPTION
**Pull request by chrisnojima**
- [x] Less calls to get user metadata
- [x] Reduce Icon rendering
- [x] We were loading messages as you click around the inbox when really we meant to load only if we hadn't loaded yet
- [x] If we've ever loaded an avatar don't use the noAvatar as a loading state as we know it'll load. Stops flickering
- [x] Make box stateless (small speedup, but everywhere)
- [x] Make usernames component have shouldComponentUpdate. Uses stringify but faster than all the rerendering we were doing. Cuts inbox rendering while clicking around in half!

### CODE REVIEW

File: shared/tab-bar/index.render.desktop.js [hunk 1]
```diff
@@ -17,6 +17,13 @@ import {
 
 import type {Props} from './index.render'
 
+const _searchSource = {type: 'nav', icon: 'iconfont-nav-search'}
+const _folderSource = {type: 'nav', icon: 'iconfont-folder'}
+const _chatSource = {type: 'nav', icon: 'iconfont-chat'}
+const _peopleSource = {type: 'nav', icon: 'iconfont-people'}
+const _devicesSource = {type: 'nav', icon: 'iconfont-device'}
+const _settingsSource = {type: 'nav', icon: 'iconfont-settings'}
```
Comment (chrisnojima) [inline shared/tab-bar/index.render.desktop.js]: Don't mutate these during render else we'll keep rerendering the same stuff

File: shared/common-adapters/icon.desktop.js [hunk 2]
```diff
@@ -9,7 +10,16 @@ import {resolveImageAsURL} from '../../desktop/resolve-root'
 import type {Exact} from '../constants/types/more'
 import type {Props} from './icon'
 
-class Icon extends PureComponent<void, Exact<Props>, void> {
+class Icon extends Component<void, Exact<Props>, void> {
+  shouldComponentUpdate (nextProps: Exact<Props>, nextState: any): boolean {
```
Comment (chrisnojima) [inline shared/common-adapters/icon.desktop.js]: Do a shallow compare, but another shallow compare on the style attribute. 99% of the time the type / style changes but ultimately is the equivalent object

File: shared/chat/conversation/messages/text.desktop.js [hunk 3]
```diff
@@ -70,6 +71,12 @@ export default class MessageTextComponent extends PureComponent<void, Props, voi
   }
 }
 
+export default withHandlers({
```
Comment (chrisnojima) [inline shared/chat/conversation/messages/text.desktop.js]: use recompose to pull out a dynamic handler so what icon gets doesn't mutate

File: shared/chat/conversation/list.desktop.js [hunk 4]
```diff
@@ -172,6 +172,10 @@ class ConversationList extends Component<void, Props, State> {
     ReactDOM.render(popupComponent, container)
   }
 
+  _onAction = (message, event) => {
```
Comment (chrisnojima) [inline shared/chat/conversation/list.desktop.js]: don't mutate this on every call to _rowRenderer

File: shared/actions/chat.js [hunk 5]
```diff
@@ -677,17 +685,17 @@ function * _updateMetadata (action: UpdateMetadata): SagaGenerator<any, any> {
...
   const inbox = yield select(_selectedInboxSelector, conversationIDKey)
-  if (inbox && !inbox.get('validated')) {
-    return
-  }
-
   if (inbox) {
     yield put({type: Constants.updateMetadata, payload: {users: inbox.get('participants').filter(p => !p.you).map(p => p.username).toArray()}})
   }
 
+  if (inbox && !inbox.get('validated')) {
```
Comment (chrisnojima) [inline shared/actions/chat.js]: let the call to updateMetadata happen anyways

File: shared/actions/chat.js [hunk 6]
```diff
@@ -472,6 +474,11 @@ function * _loadMoreMessages (): SagaGenerator<any, any> {
 
   let next
   if (oldConversationState) {
+    if (action.payload.onlyIfUnloaded && oldConversationState.get('paginationNext')) {
```
Comment (chrisnojima) [inline shared/actions/chat.js]: only load more when we intend to actually load more. else we want to load the initial screen

**Comment by chrisnojima** [comment]
@keybase/react-hackers this is ready to look at. A grab bag of perf improvements based on hot spots in timeline / react perf

**Comment by keybase-ci-visdiff** [comment]
The commits 7b65cadf7907...fbeb2299dbd7 introduce visual changes on linux. <details><summary>:mag_right: 6 changed</summary>

 * Changed: **Profile-Your_Profile_Proof_Menu** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1482115894658-7b65cadf7907-fbeb2299dbd7-linux/Profile-Your_Profile_Proof_Menu-7b65cadf7907.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1482115894658-7b65cadf7907-fbeb2299dbd7-linux/Profile-Your_Profile_Proof_Menu-fbeb2299dbd7.png)  
   Profile-Your_Profile_Proof_Menu comparison
 * Changed: **Profile-Your_Profile_Pending_Proof_Menu_HN** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1482115894658-7b65cadf7907-fbeb2299dbd7-linux/Profile-Your_Profile_Pending_Proof_Menu_HN-7b65cadf7907.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1482115894658-7b65cadf7907-fbeb2299dbd7-linux/Profile-Your_Profile_Pending_Proof_Menu_HN-fbeb2299dbd7.png)  
   Profile-Your_Profile_Pending_Proof_Menu_HN comparison
 * Changed: **Profile-Your_Profile_Pending_Proof_Menu_DNS** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1482115894658-7b65cadf7907-fbeb2299dbd7-linux/Profile-Your_Profile_Pending_Proof_Menu_DNS-7b65cadf7907.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1482115894658-7b65cadf7907-fbeb2299dbd7-linux/Profile-Your_Profile_Pending_Proof_Menu_DNS-fbeb2299dbd7.png)  
   Profile-Your_Profile_Pending_Proof_Menu_DNS comparison
 * Changed: **Profile-Your_Profile_Pending_Proof_Menu** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1482115894658-7b65cadf7907-fbeb2299dbd7-linux/Profile-Your_Profile_Pending_Proof_Menu-7b65cadf7907.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1482115894658-7b65cadf7907-fbeb2299dbd7-linux/Profile-Your_Profile_Pending_Proof_Menu-fbeb2299dbd7.png)  
   Profile-Your_Profile_Pending_Proof_Menu comparison
 * Changed: **Profile-Your_Profile_Broken_Proof_Menu** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1482115894658-7b65cadf7907-fbeb2299dbd7-linux/Profile-Your_Profile_Broken_Proof_Menu-7b65cadf7907.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1482115894658-7b65cadf7907-fbeb2299dbd7-linux/Profile-Your_Profile_Broken_Proof_Menu-fbeb2299dbd7.png)  
   Profile-Your_Profile_Broken_Proof_Menu comparison
 * Changed: **Profile-Your_Profile_Broken** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1482115894658-7b65cadf7907-fbeb2299dbd7-linux/Profile-Your_Profile_Broken-7b65cadf7907.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1482115894658-7b65cadf7907-fbeb2299dbd7-linux/Profile-Your_Profile_Broken-fbeb2299dbd7.png)  
   Profile-Your_Profile_Broken comparison
</summary>

**Comment by keybase-ci-visdiff** [comment]
The commits 7b65cadf7907...fbeb2299dbd7 introduce visual changes on linux. <details><summary>:mag_right: 6 changed</summary>

 * Changed: **Profile-Your_Profile_Proof_Menu** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1482180248452-7b65cadf7907-fbeb2299dbd7-linux/Profile-Your_Profile_Proof_Menu-7b65cadf7907.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1482180248452-7b65cadf7907-fbeb2299dbd7-linux/Profile-Your_Profile_Proof_Menu-fbeb2299dbd7.png)  
   Profile-Your_Profile_Proof_Menu comparison
 * Changed: **Profile-Your_Profile_Pending_Proof_Menu_HN** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1482180248452-7b65cadf7907-fbeb2299dbd7-linux/Profile-Your_Profile_Pending_Proof_Menu_HN-7b65cadf7907.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1482180248452-7b65cadf7907-fbeb2299dbd7-linux/Profile-Your_Profile_Pending_Proof_Menu_HN-fbeb2299dbd7.png)  
   Profile-Your_Profile_Pending_Proof_Menu_HN comparison
 * Changed: **Profile-Your_Profile_Pending_Proof_Menu_DNS** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1482180248452-7b65cadf7907-fbeb2299dbd7-linux/Profile-Your_Profile_Pending_Proof_Menu_DNS-7b65cadf7907.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1482180248452-7b65cadf7907-fbeb2299dbd7-linux/Profile-Your_Profile_Pending_Proof_Menu_DNS-fbeb2299dbd7.png)  
   Profile-Your_Profile_Pending_Proof_Menu_DNS comparison
 * Changed: **Profile-Your_Profile_Pending_Proof_Menu** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1482180248452-7b65cadf7907-fbeb2299dbd7-linux/Profile-Your_Profile_Pending_Proof_Menu-7b65cadf7907.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1482180248452-7b65cadf7907-fbeb2299dbd7-linux/Profile-Your_Profile_Pending_Proof_Menu-fbeb2299dbd7.png)  
   Profile-Your_Profile_Pending_Proof_Menu comparison
 * Changed: **Profile-Your_Profile_Broken_Proof_Menu** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1482180248452-7b65cadf7907-fbeb2299dbd7-linux/Profile-Your_Profile_Broken_Proof_Menu-7b65cadf7907.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1482180248452-7b65cadf7907-fbeb2299dbd7-linux/Profile-Your_Profile_Broken_Proof_Menu-fbeb2299dbd7.png)  
   Profile-Your_Profile_Broken_Proof_Menu comparison
 * Changed: **Profile-Your_Profile_Broken** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1482180248452-7b65cadf7907-fbeb2299dbd7-linux/Profile-Your_Profile_Broken-7b65cadf7907.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1482180248452-7b65cadf7907-fbeb2299dbd7-linux/Profile-Your_Profile_Broken-fbeb2299dbd7.png)  
   Profile-Your_Profile_Broken comparison
</summary>

**Comment by chrisnojima** [comment]
@keybase/react-hackers bump

**Comment by chrisnojima** [comment]
@keybase/react-hackers bump

File: shared/actions/chat.js [hunk 7]
```diff
@@ -760,7 +767,7 @@ function * _updateBadging (action: UpdateBadging): SagaGenerator<any, any> {
   const {conversationIDKey} = action.payload
   const conversationStateSelector = (state: TypedState) => state.chat.get('conversationStates', Map()).get(conversationIDKey)
   const conversationState = yield select(conversationStateSelector)
-  if (conversationState && conversationState.messages !== null && conversationState.messages.size > 0) {
+  if (conversationState && conversationState.firstNewMessageID && conversationState.messages !== null && conversationState.messages.size > 0) {
```
Comment (MarcoPolo) [inline shared/actions/chat.js]: What does this change do?

File: shared/chat/container.js [hunk 8]
```diff
@@ -1,23 +1,7 @@
...
-}
+import {connect} from 'react-redux'
 
 export default connect(
   (state: any) => ({}),
-  (dispatch: Dispatch) => ({
-    loadInbox: () => dispatch(loadInbox()),
-  })
-)(ChatContainer)
+)(Chat)
```
Comment (MarcoPolo) [inline shared/chat/container.js]: Does this need to be connected?

File: shared/actions/chat.js (hunk 7 above)
Comment (chrisnojima) [inline shared/actions/chat.js]: we only change the badging if we have unread stuff (and therefore the badge would change)

File: shared/chat/container.js (hunk 8 above)
Comment (chrisnojima) [inline shared/chat/container.js]: Maybe not but i didn't want to make that change until it was clear. i could imagine we might need some thing (like a modal or something) that needs it. seems harmless like this for now

**Comment by chromakode** [comment]
@chrisnojima It looks like these changes break the menu-visible profile screenshots. See the visdiff here: https://github.com/keybase/client/pull/5138

**Pull request merged**